### PR TITLE
Reimplement polynomial_regression.py

### PR DESCRIPTION
**Pull request by tianyizheng02**
### Describe your change:

Contributes to #6216

- Reimplemented [machine_learning/polynomial_regression.py](https://github.com/TheAlgorithms/Python/blob/a03b739d23b59890b59d2d2288ebaa56e3be47ce/machine_learning/polymonial_regression.py) using numpy because the old original implementation was just a how-to on doing polynomial regression using sklearn
- Added detailed function documentation, doctests, and algorithm explanation
- Fixed typo in file name: polymonial_regression.py -> polynomial_regression.py

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [x] Documentation change?

### Checklist:
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [ ] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms include at least one URL that points to Wikipedia or another similar explanation.
* [x] If this pull request resolves one or more open issues then the description above includes the issue number(s) with a [closing keyword](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue): "Fixes #ISSUE-NUMBER".

### CODE REVIEW

**Comment by tianyizheng02** [comment]
```
=================================== FAILURES ===================================
__ [doctest] machine_learning.polynomial_regression.PolynomialRegression.fit ___
133         Traceback (most recent call last):
134         ...
135         ArithmeticError: Design matrix is not full rank, can't compute coefficients
136 
137         Make sure errors don't grow too large:
138         >>> coefs = np.array([-250, 50, -2, 36, 20, -12, 10, 2, -1, -15, 1])
139         >>> y = PolynomialRegression._design_matrix(x, len(coefs) - 1) @ coefs
140         >>> poly_reg = PolynomialRegression(degree=len(coefs) - 1)
141         >>> poly_reg.fit(x, y)
142         >>> np.allclose(poly_reg.params, coefs, atol=10e-5)
Expected:
    True
Got:
    False
```
This doctest passes when I run it locally on my own machine, so I'm guessing the failure here is due to some variation in numpy's calculations on different systems

**Comment by tianyizheng02** [comment]
```
=================================== FAILURES ===================================
__ [doctest] machine_learning.polynomial_regression.PolynomialRegression.fit ___
133         Traceback (most recent call last):
134         ...
135         ArithmeticError: Design matrix is not full rank, can't compute coefficients
136 
137         Make sure errors don't grow too large:
138         >>> coefs = np.array([-250, 50, -2, 36, 20, -12, 10, 2, -1, -15, 1])
139         >>> y = PolynomialRegression._design_matrix(x, len(coefs) - 1) @ coefs
140         >>> poly_reg = PolynomialRegression(degree=len(coefs) - 1)
141         >>> poly_reg.fit(x, y)
142         >>> poly_reg.params    # Output params for debugging
Expected nothing
Got:
    array([-250.0000132 ,   50.00017691,   -2.00040817,   36.00038528,
             19.99980068,  -11.99993801,    9.99998778,    2.00000153,
             -1.00000012,  -14.99999999,    1.        ])
```
The estimated parameters clearly _are_ close to the true parameters, so I'll just set the tolerance to a larger value.

File: machine_learning/polynomial_regression.py [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+class PolynomialRegression:
+    __slots__ = "degree", "params"
+
+    def __init__(self, degree: int) -> None:
+        """
```
Comment (CaedenPH) [inline machine_learning/polynomial_regression.py:45]: This docstring should go in the base of the class, not in the __init__

File: machine_learning/polynomial_regression.py [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        @raises ValueError: if the polynomial degree is negative
+
+        >>> pass
```
Comment (CaedenPH) [inline machine_learning/polynomial_regression.py:48]: What?

File: machine_learning/polynomial_regression.py [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+                        prediction
+        @param degree:  the polynomial degree m
+        @returns:       the Vandermonde matrix X (see above)
+        @raises ValueError: if input data is not N x 1
+
+        >>> x = np.array([0, 1, 2])
+        >>> PolynomialRegression._design_matrix(x, degree=0)
+        array([[1],
+               [1],
+               [1]])
```
Comment (CaedenPH) [inline machine_learning/polynomial_regression.py:80]:
```suggestion
        array([[1], [1], [1]])
```
These arrays don't have to be broken across so many lines

File: machine_learning/polynomial_regression.py [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+
+        return PolynomialRegression._design_matrix(data, self.degree) @ self.params
+
+
+def main() -> None:
+    """
+    Fit a polynomial regression model to predict fuel efficiency using seaborn's mpg
+    dataset
+
+    >>> pass    # No doctests, function is for demo purposes only
```
Comment (CaedenPH) [inline machine_learning/polynomial_regression.py:191]: Why not just call main in the doctest?

File: machine_learning/polynomial_regression.py [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        function computes X⁺ using singular value decomposition (SVD).
+
+        References:
+            - https://en.wikipedia.org/wiki/Moore%E2%80%93Penrose_inverse
+            - https://en.wikipedia.org/wiki/Singular_value_decomposition
+            - https://en.wikipedia.org/wiki/Multicollinearity
+
+        @param x_train: the predictor values x for model fitting
+        @param y_train: the response values y for model fitting
+        @returns:       None
```
Comment (CaedenPH) [inline machine_learning/polynomial_regression.py:121]:
```suggestion
```
Doesn't seem overly helpful in my opinion

**Review by CaedenPH — COMMENTED** [review]
Very nice documentation about each function

File: machine_learning/polynomial_regression.py (hunk 3 above)
Comment (tianyizheng02) [reply machine_learning/polynomial_regression.py:80]: Doctests look for exact matches, so I'm pretty sure removing the whitespace would break this doctest

**Review by cclauss — APPROVED** [review]

**Pull request merged**